relayer/processor/path_processor.go

```diff
@@ -169,32 +180,71 @@ func (pp *PathProcessor) HandleNewData(chainID string, cacheData ChainProcessorC
 }
 
 // Run executes the main path process.
-func (pp *PathProcessor) Run(ctx context.Context) {
+func (pp *PathProcessor) Run(ctx context.Context, ctxCancel func(), messageLifecycle MessageLifecycle) {
+	var retryTimer *time.Timer
 	for {
+		// block until we have any signals to process
 		select {
 		case <-ctx.Done():
+			pp.log.Debug("Context done, quitting PathProcessor",
+				zap.String("chain_id_1", pp.pathEnd1.info.ChainID),
+				zap.String("chain_id_2", pp.pathEnd2.info.ChainID),
+				zap.String("client_id_1", pp.pathEnd1.info.ClientID),
+				zap.String("client_id_2", pp.pathEnd2.info.ClientID),
+				zap.Error(ctx.Err()),
+			)
 			return
-
 		case d := <-pp.pathEnd1.incomingCacheData:
 			// we have new data from ChainProcessor for pathEnd1
-			pp.pathEnd1.MergeCacheData(d)
+			pp.pathEnd1.MergeCacheData(ctx, ctxCancel, d, messageLifecycle)
 
 		case d := <-pp.pathEnd2.incomingCacheData:
 			// we have new data from ChainProcessor for pathEnd2
-			pp.pathEnd2.MergeCacheData(d)
+			pp.pathEnd2.MergeCacheData(ctx, ctxCancel, d, messageLifecycle)
 
 		case <-pp.retryProcess:
 			// No new data to merge in, just retry handling.
 		}
 
+		// Fully flush pathEnd incoming data before processing
+		for len(pp.pathEnd1.incomingCacheData) > 0 {
+			pp.pathEnd1.MergeCacheData(ctx, ctxCancel, <-pp.pathEnd1.incomingCacheData, messageLifecycle)
+		}
+		for len(pp.pathEnd2.incomingCacheData) > 0 {
+			pp.pathEnd2.MergeCacheData(ctx, ctxCancel, <-pp.pathEnd2.incomingCacheData, messageLifecycle)
+		}
+
+		// flush retry process in case retries were scheduled
+		for len(pp.retryProcess) > 0 {
+			<-pp.retryProcess
```

Comment:
This also looks like it could be a select against this channel and `ctx.Done`. Unlikely to hit the `ctx.Done` if everything works as intended, but (somewhat) blocks against a potential programming bug where retries come in infinitely.

Reply:
Yes I don't really see any cases where it would be > 0, but I didn't want unnecessary retries to keep triggering the `select` above. I think I would rather remove this flush and just have the one up in the `select` than have something that pulls an additional retry off the channel. Maybe I will just reduce the size of the channel to `1`.

Reply:
Another reason I like this flush though is because it ensures that if both `pp.pathEnd1.incomingCacheData` and `pp.retryProcess` have an item, and the `select` picks the former, the `retryProcess` is also guaranteed to be cleared since we don't need an additional retry if the `processLatestMessages` is ran for the new `incomingCacheData`. If we block for either `<-ctx.Done()` or `<-pp.retryProcess` down here again, it will hang during normal operation.